src/Internal/Plutip/Server.purs

```diff
@@ -124,24 +139,97 @@ withPlutipContractEnv
   -> (ContractEnv () -> wallets -> Aff a)
   -> Aff a
 withPlutipContractEnv plutipCfg distr cont = do
+  cleanupRef <- liftEffect $ Ref.new mempty
+  Aff.bracket
+    (startPlutipContractEnv plutipCfg distr cleanupRef)
+    (const $ cleanup cleanupRef)
+    \{ env, wallets, printLogs } -> do
+      runWithLogs printLogs (cont env wallets)
+
+-- | Run `Contract`s in tests in a single Plutip instance.
+testPlutipContracts
+  :: forall (distr :: Type) (wallets :: Type)
+   . UtxoDistribution distr wallets
+  => PlutipConfig
+  -> distr
+  -> TestPlanM (wallets -> Contract () Unit) Unit
+  -> TestPlanM (Aff Unit) Unit
+testPlutipContracts plutipCfg distr tests = do
+  let groupedTests = group "HIDDEN_GROUP" tests
+  cleanupRef <- liftEffect $ Ref.new mempty
+  bracket (startPlutipContractEnv plutipCfg distr cleanupRef)
+    (cleanup cleanupRef)
+    $ flip mapTest groupedTests \test { env, wallets, printLogs } -> do
+        runWithLogs printLogs (runContractInEnv env (test wallets))
+  where
+  bracket
+    :: forall a b
+     . Aff a
+    -> Aff Unit
+    -> TestPlanM (a -> Aff b) Unit
+    -> TestPlanM (Aff b) Unit
+  bracket before' after' act = do
+    resultRef <- liftEffect $ Ref.new (Left $ error "Plutip not initialized")
+    let
+      before = do
+        res <- try $ before'
+        liftEffect $ Ref.write res resultRef
+        pure res
+      after = const $ after'
+    Mote.bracket { before, after } $ flip mapTest act \t -> do
+      result <- liftEffect $ Ref.read resultRef >>= liftEither
+      t result
+
+-- | Provide a `ContractEnv` connected to Plutip.
+-- | can be used to run multiple `Contract`s using `runContractInEnv`.
+startPlutipContractEnv
+  :: forall (distr :: Type) (wallets :: Type)
+   . UtxoDistribution distr wallets
+  => PlutipConfig
+  -> distr
+  -> Ref (Array (Aff Unit))
+  -> Aff
+       { env :: ContractEnv ()
+       , wallets :: wallets
+       , printLogs :: Aff Unit
+       }
+startPlutipContractEnv plutipCfg distr cleanupRef = do
   configCheck plutipCfg
-  withPlutipServer $
-    withPlutipCluster \(ourKey /\ response) ->
-      withPostgres response
-        $ withOgmios response
-        $ withOgmiosDatumCache response
-        $ withMCtlServer
-        $ withContractEnv \env ->
-            withWallets env ourKey response (cont env)
+  startPlutipServer'
+  ourKey /\ response <- startPlutipCluster'
+  startPostgres'
```

Comment:
I don't get what's happening here. Has the bracketing been removed?
Judging by the code, you are starting services and then letting them immediately stop.

Reply:
No, bracketting is still there, just achieved differently to be compatible with test brackets. See the bracket function these starters use in the where

Reply:
OK I see, please explain the motivation in a comment.